src/Elm/Compiler/Error.elm

```diff
@@ -198,10 +202,56 @@ toString error =
                     "Invalid elm.json! "
                         ++ JD.errorToString jsonError
 
-                ParseProblem problems ->
-                    String.join "\n"
-                        ("Parse problems: "
-                            :: List.map (\{ problem } -> "  " ++ parseProblemToString problem) problems
+                ParseProblem ( problems, source ) ->
+                    String.join
+                        "\n"
+                        (List.map
+                            (\{ problem, row, col, contextStack } ->
+                                let
+                                    filename =
+                                        contextStack
+                                            |> List.filterMap
+                                                (\{ context } ->
+                                                    case context of
+                                                        InFile name ->
+                                                            Just name
+
+                                                        _ ->
+                                                            Nothing
+                                                )
+                                            |> List.Extra.last
```

Comment:
The way this is set up, do we ever get more than one `InFile`? Couldn't we just use `List.Extra.find`?

Reply:
Yes, I think we could. With fix

Reply:
I am having second thoughts, currently we have:

```elm
filename =
    contextStack
        |> List.Extra.filterMap
            (\{ context } ->
                case context of
                    InFile name ->
                        Just name

                    _ ->
                        Nothing
            )
        |> List.Extra.last
````

I cannot think of a better way to write it using `List.Extra.find` than

```elm
filename =
    contextStack
        |> List.Extra.map
            (\{ context } ->
                case context of
                    InFile name ->
                        Just name

                    _ ->
                        Nothing
            )
        |> List.Extra.find 
            (\{ name_ } ->
                case name_ of
                    Just name ->
                        True

                    _ ->
                        False
            )
        |> Maybe.withDefault Nothing
```

which is horrible!

Thoughts?

Reply:
True, when I tried, I got to similar issues. So what about custom recursion?
```elm
filename =
    findFirstJustcontextStack
    contextStack
        |> List.filterMap
            (\{ context } ->
                case c
```

Reply:
True, when I tried I got into similar issues. So what about custom recursion? (Maybe this has some List.Extra equivalent...)
```elm
filename =
    findFirstJust 
        (\{ context } ->
            case context of
                InFile name ->
                    Just name
                _ ->
                    Nothing
        )
        contextStack

findFirstJust : (a -> Maybe b) -> List a -> Maybe b
findFirstJust fn list =
    case list of
        [] ->
            Nothing

        x :: xs ->
            case fn x of
                Just result ->
                    Just result

                Nothing ->
                    findFirstJust fn xs
```

Reply:
(Or let's keep the `filterMap` and `last` / `first` / ..., considering the alternatives aren't much better after all)

Reply:
How about:

```elm
filenameFromContext : List ParseContext -> Maybe FilePath
filenameFromContext contextList =
    case contextList of
        context :: rest ->
            case context of
                InFile name ->
                    Just name
                _ ->
                    filenameFromContext rest

        [] ->
            Nothing
```

Reply:
Looks good!

Reply:
will sort next time I am on my laptop 👍